src/Switch-FTDI_FTD2XX/main.py

```diff
@@ -0,0 +1,143 @@
+
+import os
+import sys
+
+file_dir = os.path.dirname(__file__)
+sys.path.append(file_dir)
+
+libs_path = file_dir + os.sep + "libs"
+libraries_path = file_dir + os.sep + r"libraries\libs_39_64"
+sys.path.append(libraries_path)
+
+
+import binascii
+
+import ftd2xx as ftd
+from pysweepme import EmptyDevice
+
+
+class Device(EmptyDevice):
+    description = """
+        Need FTDI driver installed
+        Can send initial command when initializing or end command when deinitializing
+    """
+
+    def __init__(self) -> None:
+        EmptyDevice.__init__(self)
+
+        self.shortname = "FTDI245"
+        self.variables = []
+        self.units = []
+
+        self.port_manager = False
+        self.port_types = ["USB"]
+
+    def find_ports(self):
+        # print(ftd.listDevices())
+        return [dev.decode() for dev in ftd.listDevices()]
+
+    def set_GUIparameter(self):
+        return {
+            "Command": "Some Command",
+            "Encoding": ["HEX", "ASCII"],
+            "Init Command": "",
+            "End Command": "",
+        }
+
+    def get_GUIparameter(self, parameter={}):
+        self.encoding = parameter["Encoding"]
+
+        self.port_str = parameter["Port"]
+        self.driver_name = parameter["Device"]
+
+    """ here, semantic standard functions start that are called by SweepMe! during a measurement """
+
+    def connect(self):
+        # Check if port string is still in list of devices - maybe double?
+        if self.port_str not in ftd.listDevices():
+            msg = f"FTD Device with serial number {self.port_str} not found"
+            raise Exception(msg)
+
+        # Open device
+        try:
+            self.device = ftd.openEx(self.port_str)
+        except ftd.ftd2xx.DeviceError:
+            msg = f"Cannot open FTD Device with serial number {self.port_str}"
+            raise Exception(msg)
+
+
+        # Set Name/Number of COM Port as key
+        self.instance_key = f"{self.driver_name}_{self.port_str}"
+
+        if self.instance_key not in self.device_communication:
+            self.device_communication[self.instance_key] = "Connected"
+
+    def disconnect(self):
+        # Close device
+        self.device.close()
+
+        if self.instance_key in self.device_communication:
+            self.device_communication.pop(self.instance_key)
+
+    def configure(self):
+        self.device.setTimeouts(5000, 5000)
+
+    def unconfigure(self):
+        # TODO: Timeout?
+        pass
+
+    def apply(self):
+        # command: 5 Bytes
+
+        if self.encoding == "HEX":
+            # length 5
+            command = binascii.unhexlify(self.value.replace(" ", ""))
+        # elif self.encoding == "ASCII":
+        else:
+            # length 14
+            command = bytes(self.value, "utf-8")
+
+        self.device.write(command)
+
+    def call(self):
+        # Number of bytes in queue
+        queue = self.device.getQueueStatus()
+
+        # check timeout?
+
+        if queue > 0:
```

Comment:
If you like to return a message string, it must be done in any case. The code works so far because the case where there are some characters in the buffer did not happen so far.
If a value is returned in 'call', self.variables, self.units, etc. must be adapted accordingly.